synapseclient/core/download/download_functions.py

```diff
@@ -719,21 +735,64 @@ def _ftp_report_hook(
                 else None
             )
 
-            response = with_retry(
-                lambda url=url, range_header=range_header, auth=auth: client._requests_session.get(
-                    url=url,
-                    headers=client._generate_headers(range_header),
-                    stream=True,
-                    allow_redirects=False,
-                    auth=auth,
-                ),
-                verbose=client.debug,
-                **STANDARD_RETRY_PARAMS,
-            )
             try:
+                url_has_expiration = "Expires" in urllib_urlparse.urlparse(url).query
+                url_is_expired = False
+                if url_has_expiration:
+                    url_is_expired = datetime.datetime.now(
+                        tz=datetime.timezone.utc
+                    ) + PresignedUrlProvider._TIME_BUFFER >= _pre_signed_url_expiration_time(
+                        url
+                    )
+                if url_is_expired:
+                    response = get_file_handle_for_download(
+                        file_handle_id=file_handle_id,
+                        synapse_id=entity_id,
+                        entity_type=file_handle_associate_type,
+                        synapse_client=client,
+                    )
+                    url = response["preSignedURL"]
+                response = with_retry(
+                    lambda url=url, range_header=range_header, auth=auth: client._requests_session.get(
+                        url=url,
+                        headers=client._generate_headers(range_header),
+                        stream=True,
+                        allow_redirects=False,
+                        auth=auth,
+                    ),
+                    verbose=client.debug,
+                    **STANDARD_RETRY_PARAMS,
+                )
                 exceptions._raise_for_status(response, verbose=client.debug)
             except SynapseHTTPError as err:
-                if err.response.status_code == 404:
+                if err.response.status_code == 403:
+                    url_is_expired = datetime.datetime.now(
```

Comment:
Does the additional check you added:

```
                url_has_expiration = "Expires" in urllib_urlparse.urlparse(url).query
                url_is_expired = False
                if url_has_expiration:
```

Need to also be added here?

Reply:
Is it possible for the URL to expire between the first and second download attempts?

Reply:
> Is it possible for the URL to expire between the first and second download attempts?

Likely no. However from what you had mentioned:

> URL that does not have an expiration time and so this errors out if we add this check before the initial attempt.

If we have a URL that does not have an expiration time AND it got an HTTP 403 error back. Wouldn't that raise an exception here since the code is expecting that the URL has a expiration time on it?

Reply:
OK I added this and a test for the case when a URL does not have an expiration datetime.